src/ext/frame-multiplexer.lisp

```diff
@@ -283,38 +330,98 @@
+  "Switch to the Nth previous frame.
+The prefix argument N defaults to 1."
   (check-frame-multiplexer-enabled)
   (let* ((vf (gethash (implementation) *virtual-frame-map*))
-         (frame (search-previous-frame vf (virtual-frame-current vf))))
-    (when frame
-      (switch-current-frame vf frame))))
+         (frame (virtual-frame-current vf)))
+    (dotimes (_ n)
+      (setq frame (search-previous-frame vf frame)))
+    (switch-current-frame vf frame)))
 
 (define-key *global-keymap* "C-z n" 'frame-multiplexer-next)
-(define-command frame-multiplexer-next () ()
+(define-command frame-multiplexer-next (&optional (n 1)) ("p")
+  "Switch to the Nth next frame.
+The prefix argument N defaults to 1."
   (check-frame-multiplexer-enabled)
   (let* ((vf (gethash (implementation) *virtual-frame-map*))
-         (frame (search-next-frame vf (virtual-frame-current vf))))
-    (when frame
-      (switch-current-frame vf frame))))
+         (frame (virtual-frame-current vf)))
+    (dotimes (_ n)
+      (setq frame (search-next-frame vf frame)))
+    (switch-current-frame vf frame)))
+
+(define-command frame-multiplexer-switch (&optional (id 1)) ("p")
+  "Switch to the frame with ID.
+The prefix argument ID defaults to 1."
+  ;; TODO: It would be great to enhance this by showing a prompt
+  ;; and asking for the frame name (or buffer of the frame, if it has no name)
+  (check-frame-multiplexer-enabled)
+  (let* ((vf (gethash (implementation) *virtual-frame-map*))
+         (entry (aref (virtual-frame-id/frame-table vf) id)))
+    (if entry
+        (switch-current-frame vf (frame-table-entry-frame entry))
+        (editor-error "No frame with ID ~a" id))))
+
+(define-command frame-multiplexer-recent (&optional (n 1)) ("p")
+  "Switch to the Nth most recent frame selected.
+The prefix argument N defaults to 1."
+  (check-frame-multiplexer-enabled)
+  (unless (or (ring-empty-p *recent-list*)
+              (>= 0 n))
+    (let* ((vf (gethash (implementation) *virtual-frame-map*))
+           ;; Loop over the ring and skip IDs which are no longer valid.
+           (recent-frame-id (loop :with valid-count := 0
+                                  :for ref-n :upfrom 0
+                                  :when (>= ref-n (ring-length *recent-list*))
+                                    :return nil
+                                  :when (aref (virtual-frame-id/frame-table vf) (ring-ref *recent-list* ref-n))
+                                    :do (incf valid-count)
+                                  :when (= valid-count n)
+                                    :return (ring-ref *recent-list* ref-n))))
+      (if (null recent-frame-id)
+          (editor-error "No more recent frames")
+          (let ((entry (aref (virtual-frame-id/frame-table vf) recent-frame-id)))
+            (switch-current-frame vf (frame-table-entry-frame entry)))))))
+
+(define-key *global-keymap* "C-z r" 'frame-mulitplexer-rename)
```

Comment:
I think the current style is to put all the `define-key`s at the top of the file, though there is no official rule and I don't know the maintainer's opinion.

Reply:
Indeed, I think that is better.

Reply:
Fixed, moved all `define-key`s to the top of the file.